src/core/lib/support/function.h

```diff
@@ -0,0 +1,293 @@
+ public:
+  template <class T>
+  static const VTable<R, Args...>* Construct(F&& f, T* storage) {
+    static_assert(sizeof(*storage) >= sizeof(F), "Size of storage too small");
+    new (storage) F(std::move(f));
+    return &vtable_;
+  }
+
+ private:
+  static void CopyConstruct(const void* from, void* to) {
+    new (to) F(*static_cast<const F*>(from));
+  }
+  static void MoveConstruct(void* from, void* to) {
+    new (to) F(std::move(*static_cast<F*>(from)));
+  }
+  static void Destruct(void* storage) { static_cast<F*>(storage)->~F(); }
+  static R Invoke(void* storage, Args&&... args) {
+    return (*static_cast<F*>(storage))(std::forward<Args>(args)...);
+  }
+
+  static const VTable<R, Args...> vtable_;
+};
+
+template <typename R, typename... Args, typename F>
+const VTable<R, Args...> Traits<R(Args...), F, false>::vtable_ = {
+    CopyConstruct, MoveConstruct, Destruct, Invoke};
+
+template <typename R, typename... Args, typename F>
+class Traits<R(Args...), F, true> {
+ private:
+  typedef UniquePtr<F> Ptr;
+
+ public:
+  static const VTable<R, Args...>* Construct(F&& f, void* storage) {
+    new (storage) Ptr(New<F>(std::move(f)));
+    return &vtable_;
+  }
+
+ private:
+  static void CopyConstruct(const void* from, void* to) {
+    new (to) Ptr(New<F>(*static_cast<const Ptr*>(from)->get()));
+  }
+  static void MoveConstruct(void* from, void* to) {
+    new (to) Ptr(std::move(*static_cast<Ptr*>(from)));
+  }
+  static void Destruct(void* storage) { static_cast<Ptr*>(storage)->~Ptr(); }
+  static R Invoke(void* storage, Args&&... args) {
+    return (**static_cast<Ptr*>(storage))(std::forward<Args>(args)...);
+  }
+
+  static const VTable<R, Args...> vtable_;
+};
+
+template <typename R, typename... Args, typename F>
+const VTable<R, Args...> Traits<R(Args...), F, true>::vtable_ = {
+    CopyConstruct, MoveConstruct, Destruct, Invoke};
+}
+
+template <typename T,
+          size_t kInplaceStorage = function_impl::kDefaultInplaceStorage>
+class Function;
+
+template <typename T,
+          size_t kInplaceStorage = function_impl::kDefaultInplaceStorage>
+class InplaceFunction;
+
+template <typename T,
+          size_t kInplaceStorage = function_impl::kDefaultInplaceStorage>
+class TrivialInplaceFunction;
+
+// Function<> is roughly a std::function work-alike
+// Differences:
+// - configurable inplace storage quantity... functors smaller than this incur
+//   no allocation overhead, functors bigger will cause an allocation
+template <typename R, typename... Args, size_t kInplaceStorage>
+class Function<R(Args...), kInplaceStorage> {
+ public:
+  template <class F>
+  Function(F&& f) {
+    vtable_ =
+        function_impl::Traits<R(Args...), F, (sizeof(F) > kInplaceStorage)>::
+            Construct(std::forward<F>(f), &storage_);
+  }
+
+  Function(const Function& rhs) : vtable_(rhs.vtable_) {
+    vtable_->copy_construct(&rhs.storage_, &storage_);
+  }
+
+  Function(Function& rhs) : vtable_(rhs.vtable_) {
```

Comment:
I'm confused as to why this variant is needed. Is this for future expansion?

Reply:
I had some compilation woes without it, and it seemed harmless to add.

Reply:
Oh... so the actual problem:

If I call `Function()` with a `Function&` and don't have this constructor, then this one matches best:
```
template <class F>
Function(F&& f)
```

which ends up wrapping a Function around the Function and is absolutely not what we want